lib/hiera/backend.rb

```diff
@@ -233,6 +249,16 @@ def lookup(key, default, scope, order_override, resolution_type)
       def clear!
         @backends = {}
       end
+
+      def qualified_lookup(segments, hash)
+        value = hash
+        segments.each do |segment|
+          break if value.nil? || value == :undefined
```

Comment:
I think the use of `:undefined` is a relic stemming from some very old version of puppet. Since the new release is tied to Puppet 4.0 and it always uses nil as undefined when looking up in a scope, the check against :undefined can be dropped (unless it is an internal hiera marker of undefinedness).

If so, we never ever want a value of :undefined to be returned to puppet